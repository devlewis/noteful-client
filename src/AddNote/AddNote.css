.AddNote {
  margin-left: 10px;
  color: white;
}

#name {
  margin-bottom: 5px;
}

#content {
  margin: 0 10px 10px 0;
  width: 90%;
  height: 50px;
  resize: none;
}

select {
  width: 200px;
  font-size: 0.8em;
}

.AddNote_buttons {
  margin-top: 10px;
}

button {
  font-size: 0.8em;
  margin-right: 10px;
}
