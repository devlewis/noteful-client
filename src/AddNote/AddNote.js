import React from "react";
import config from "../config";
import "./AddNote.css";
import Context from "../Context";
import ValidationError from "../ValidationError";

class AddNote extends React.Component {
  static defaultProps = {};

  static contextType = Context;

  constructor(props) {
    super(props);
    this.state = {
      error: null,
      folderNameChoice: "",
      name: {
        value: "",
        touched: false
      },
      content: ""
    };
  }

  validateName() {
    const name = this.state.name.value.trim();
    if (name.length === 0) {
      return "Note name is required";
    } else if (name.length < 1) {
      return "Note name must be more than 1 character";
    }
  }

  handleSubmit = e => {
    e.preventDefault();

    const folderNameChoice = this.state.folderNameChoice;

    const folder = this.context.folders.find(
      ({ name }) => name === folderNameChoice
    );

    let noteNew = {
      name: this.state.name.value,
      content: this.state.content,
      folderId: folder.id
    };

    fetch(`${config.API_ENDPOINT}/notes`, {
      method: "POST",
      body: JSON.stringify(noteNew),
      headers: { "content-type": "application/json" }
    })
      .then(res => {
        if (!res.ok) {
          return res.json().then(error => {
            throw error;
          });
        }
        return res.json();
      })
      .then(data => {
        this.context.addNote(data);
        this.props.history.push("/");
      })
      .catch(error => {
        this.setState({ error: error });
      });
  };

  handleClickCancel = () => {
    this.props.history.push("/");
  };

  render() {
    const nameError = this.validateName();

    const { error } = this.state;
    this.context;
    const folderNames = this.context.folders
      ? this.context.folders.map(folder => (
          <option key={folder.id} value={folder.name}>
            {folder.name}
          </option>
        ))
      : [];

    return (
      <section className="AddNote">
        <h2>Create a Note</h2>
        <form className="AddNote_form" onSubmit={this.handleSubmit}>
          <div className="AddNote_error" role="alert">
            {error && <p>{error.message}</p>}
          </div>
          <div>
            <label htmlFor="name">Note Name </label>
            <input
              onChange={e =>
                this.setState({
                  name: {
                    value: e.target.value,
                    touched: true
                  }
                })
              }
              type="text"
              name="name"
              id="name"
              placeholder="name"
              autocomplete="off"
              required
            />
            {this.state.name.touched && <ValidationError message={nameError} />}
          </div>
          <div>
            <label htmlFor="content">Content </label>
            <textarea
              onChange={e => this.setState({ content: e.target.value })}
              type="text"
              name="content"
              id="content"
              autocomplete="off"
            />
          </div>
          <div>
            <label htmlFor="folderNameChoice">Folder </label>
            <select
              onChange={e =>
                this.setState({ folderNameChoice: e.target.value })
              }
              name="folderNameChoice"
              id="folderNameChoice"
              required
            >
              <option value="">Select a Folder</option>
              {folderNames}
            </select>
          </div>
          <div className="AddNote_buttons">
            <button type="button" onClick={this.handleClickCancel}>
              Cancel
            </button>
            {""}
            <button type="submit">Save</button>
          </div>
        </form>
      </section>
    );
  }
}

export default AddNote;
