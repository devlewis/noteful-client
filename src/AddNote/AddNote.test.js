import React from "react";
import ReactDOM from "react-dom";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import AddNote from "./AddNote";

describe(`AddNote component`, () => {
  it("renders without errors", () => {
    expect(shallow(<AddNote />)).toMatchSnapshot();
  });
});
