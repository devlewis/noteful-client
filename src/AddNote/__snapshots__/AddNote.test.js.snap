// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote component renders without errors 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddNote />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Create a Note
        </h2>,
        <form
          className="AddNote_form"
          onSubmit={[Function]}
        >
          <div
            className="AddNote_error"
            role="alert"
          />
          <div>
            <label
              htmlFor="name"
            >
              Note Name 
            </label>
            <input
              autocomplete="off"
              id="name"
              name="name"
              onChange={[Function]}
              placeholder="name"
              required={true}
              type="text"
            />
          </div>
          <div>
            <label
              htmlFor="content"
            >
              Content 
            </label>
            <textarea
              autocomplete="off"
              id="content"
              name="content"
              onChange={[Function]}
              type="text"
            />
          </div>
          <div>
            <label
              htmlFor="folderNameChoice"
            >
              Folder 
            </label>
            <select
              id="folderNameChoice"
              name="folderNameChoice"
              onChange={[Function]}
              required={true}
            >
              <option
                value=""
              >
                Select a Folder
              </option>
            </select>
          </div>
          <div
            className="AddNote_buttons"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
            
            <button
              type="submit"
            >
              Save
            </button>
          </div>
        </form>,
      ],
      "className": "AddNote",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Create a Note",
        },
        "ref": null,
        "rendered": "Create a Note",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="AddNote_error"
              role="alert"
            />,
            <div>
              <label
                htmlFor="name"
              >
                Note Name 
              </label>
              <input
                autocomplete="off"
                id="name"
                name="name"
                onChange={[Function]}
                placeholder="name"
                required={true}
                type="text"
              />
            </div>,
            <div>
              <label
                htmlFor="content"
              >
                Content 
              </label>
              <textarea
                autocomplete="off"
                id="content"
                name="content"
                onChange={[Function]}
                type="text"
              />
            </div>,
            <div>
              <label
                htmlFor="folderNameChoice"
              >
                Folder 
              </label>
              <select
                id="folderNameChoice"
                name="folderNameChoice"
                onChange={[Function]}
                required={true}
              >
                <option
                  value=""
                >
                  Select a Folder
                </option>
              </select>
            </div>,
            <div
              className="AddNote_buttons"
            >
              <button
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
              
              <button
                type="submit"
              >
                Save
              </button>
            </div>,
          ],
          "className": "AddNote_form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": null,
              "className": "AddNote_error",
              "role": "alert",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  htmlFor="name"
                >
                  Note Name 
                </label>,
                <input
                  autocomplete="off"
                  id="name"
                  name="name"
                  onChange={[Function]}
                  placeholder="name"
                  required={true}
                  type="text"
                />,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Note Name ",
                  "htmlFor": "name",
                },
                "ref": null,
                "rendered": "Note Name ",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "autocomplete": "off",
                  "id": "name",
                  "name": "name",
                  "onChange": [Function],
                  "placeholder": "name",
                  "required": true,
                  "type": "text",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  htmlFor="content"
                >
                  Content 
                </label>,
                <textarea
                  autocomplete="off"
                  id="content"
                  name="content"
                  onChange={[Function]}
                  type="text"
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Content ",
                  "htmlFor": "content",
                },
                "ref": null,
                "rendered": "Content ",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "autocomplete": "off",
                  "id": "content",
                  "name": "content",
                  "onChange": [Function],
                  "type": "text",
                },
                "ref": null,
                "rendered": null,
                "type": "textarea",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  htmlFor="folderNameChoice"
                >
                  Folder 
                </label>,
                <select
                  id="folderNameChoice"
                  name="folderNameChoice"
                  onChange={[Function]}
                  required={true}
                >
                  <option
                    value=""
                  >
                    Select a Folder
                  </option>
                </select>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Folder ",
                  "htmlFor": "folderNameChoice",
                },
                "ref": null,
                "rendered": "Folder ",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <option
                      value=""
                    >
                      Select a Folder
                    </option>,
                    Array [],
                  ],
                  "id": "folderNameChoice",
                  "name": "folderNameChoice",
                  "onChange": [Function],
                  "required": true,
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Select a Folder",
                      "value": "",
                    },
                    "ref": null,
                    "rendered": "Select a Folder",
                    "type": "option",
                  },
                ],
                "type": "select",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <button
                  onClick={[Function]}
                  type="button"
                >
                  Cancel
                </button>,
                "",
                <button
                  type="submit"
                >
                  Save
                </button>,
              ],
              "className": "AddNote_buttons",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Cancel",
                  "onClick": [Function],
                  "type": "button",
                },
                "ref": null,
                "rendered": "Cancel",
                "type": "button",
              },
              "",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Save",
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Save",
                "type": "button",
              },
            ],
            "type": "div",
          },
        ],
        "type": "form",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Create a Note
          </h2>,
          <form
            className="AddNote_form"
            onSubmit={[Function]}
          >
            <div
              className="AddNote_error"
              role="alert"
            />
            <div>
              <label
                htmlFor="name"
              >
                Note Name 
              </label>
              <input
                autocomplete="off"
                id="name"
                name="name"
                onChange={[Function]}
                placeholder="name"
                required={true}
                type="text"
              />
            </div>
            <div>
              <label
                htmlFor="content"
              >
                Content 
              </label>
              <textarea
                autocomplete="off"
                id="content"
                name="content"
                onChange={[Function]}
                type="text"
              />
            </div>
            <div>
              <label
                htmlFor="folderNameChoice"
              >
                Folder 
              </label>
              <select
                id="folderNameChoice"
                name="folderNameChoice"
                onChange={[Function]}
                required={true}
              >
                <option
                  value=""
                >
                  Select a Folder
                </option>
              </select>
            </div>
            <div
              className="AddNote_buttons"
            >
              <button
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
              
              <button
                type="submit"
              >
                Save
              </button>
            </div>
          </form>,
        ],
        "className": "AddNote",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Create a Note",
          },
          "ref": null,
          "rendered": "Create a Note",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="AddNote_error"
                role="alert"
              />,
              <div>
                <label
                  htmlFor="name"
                >
                  Note Name 
                </label>
                <input
                  autocomplete="off"
                  id="name"
                  name="name"
                  onChange={[Function]}
                  placeholder="name"
                  required={true}
                  type="text"
                />
              </div>,
              <div>
                <label
                  htmlFor="content"
                >
                  Content 
                </label>
                <textarea
                  autocomplete="off"
                  id="content"
                  name="content"
                  onChange={[Function]}
                  type="text"
                />
              </div>,
              <div>
                <label
                  htmlFor="folderNameChoice"
                >
                  Folder 
                </label>
                <select
                  id="folderNameChoice"
                  name="folderNameChoice"
                  onChange={[Function]}
                  required={true}
                >
                  <option
                    value=""
                  >
                    Select a Folder
                  </option>
                </select>
              </div>,
              <div
                className="AddNote_buttons"
              >
                <button
                  onClick={[Function]}
                  type="button"
                >
                  Cancel
                </button>
                
                <button
                  type="submit"
                >
                  Save
                </button>
              </div>,
            ],
            "className": "AddNote_form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": null,
                "className": "AddNote_error",
                "role": "alert",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    htmlFor="name"
                  >
                    Note Name 
                  </label>,
                  <input
                    autocomplete="off"
                    id="name"
                    name="name"
                    onChange={[Function]}
                    placeholder="name"
                    required={true}
                    type="text"
                  />,
                  false,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Note Name ",
                    "htmlFor": "name",
                  },
                  "ref": null,
                  "rendered": "Note Name ",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "autocomplete": "off",
                    "id": "name",
                    "name": "name",
                    "onChange": [Function],
                    "placeholder": "name",
                    "required": true,
                    "type": "text",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    htmlFor="content"
                  >
                    Content 
                  </label>,
                  <textarea
                    autocomplete="off"
                    id="content"
                    name="content"
                    onChange={[Function]}
                    type="text"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Content ",
                    "htmlFor": "content",
                  },
                  "ref": null,
                  "rendered": "Content ",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "autocomplete": "off",
                    "id": "content",
                    "name": "content",
                    "onChange": [Function],
                    "type": "text",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "textarea",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    htmlFor="folderNameChoice"
                  >
                    Folder 
                  </label>,
                  <select
                    id="folderNameChoice"
                    name="folderNameChoice"
                    onChange={[Function]}
                    required={true}
                  >
                    <option
                      value=""
                    >
                      Select a Folder
                    </option>
                  </select>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Folder ",
                    "htmlFor": "folderNameChoice",
                  },
                  "ref": null,
                  "rendered": "Folder ",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <option
                        value=""
                      >
                        Select a Folder
                      </option>,
                      Array [],
                    ],
                    "id": "folderNameChoice",
                    "name": "folderNameChoice",
                    "onChange": [Function],
                    "required": true,
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Select a Folder",
                        "value": "",
                      },
                      "ref": null,
                      "rendered": "Select a Folder",
                      "type": "option",
                    },
                  ],
                  "type": "select",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <button
                    onClick={[Function]}
                    type="button"
                  >
                    Cancel
                  </button>,
                  "",
                  <button
                    type="submit"
                  >
                    Save
                  </button>,
                ],
                "className": "AddNote_buttons",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Cancel",
                    "onClick": [Function],
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": "Cancel",
                  "type": "button",
                },
                "",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Save",
                    "type": "submit",
                  },
                  "ref": null,
                  "rendered": "Save",
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "form",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
